packages/git/src/browser/git-decoration-provider.ts

```diff
@@ -14,54 +14,103 @@
+
     private readonly onDidChangeDecorationsEmitter = new Emitter<URI[]>();
     readonly onDidChange: Event<URI[]> = this.onDidChangeDecorationsEmitter.event;
 
-    private decorations = new Map<string, Decoration>();
+    @postConstruct()
+    protected init(): void {
+        this.decorationsEnabled = this.preferences['git.decorations.enabled'];
+        this.colorsEnabled = this.preferences['git.decorations.colors'];
+        this.gitRepositoryTracker.onGitEvent((event: GitStatusChangeEvent | undefined) => this.handleGitEvent(event));
+        this.preferences.onPreferenceChanged(event => this.handlePreferenceChange(event));
+    }
 
-    constructor(@inject(GitRepositoryTracker) protected readonly gitRepositoryTracker: GitRepositoryTracker) {
-        this.gitRepositoryTracker.onGitEvent((event: GitStatusChangeEvent | undefined) => {
-            this.onGitEvent(event);
-        });
+    protected async handleGitEvent(event: GitStatusChangeEvent | undefined): Promise<void> {
+        this.changeEvent = event;
+        this.updateDecorations();
     }
 
-    private async onGitEvent(event: GitStatusChangeEvent | undefined): Promise<void> {
-        if (!event) {
+    protected updateDecorations(): void {
+        if (!this.changeEvent) {
             return;
         }
-
         const newDecorations = new Map<string, Decoration>();
-        this.collectDecorationData(event.status.changes, newDecorations);
+        this.collectDecorationData(this.changeEvent.status.changes, newDecorations);
 
-        const uris = new Set([...this.decorations.keys()].concat([...newDecorations.keys()]));
+        this.uris = new Set([...this.decorations.keys()].concat([...newDecorations.keys()]));
         this.decorations = newDecorations;
-        this.onDidChangeDecorationsEmitter.fire(Array.from(uris, value => new URI(value)));
+        this.triggerDecorationChange();
     }
 
-    private collectDecorationData(changes: GitFileChange[], bucket: Map<string, Decoration>): void {
+    protected collectDecorationData(changes: GitFileChange[], bucket: Map<string, Decoration>): void {
         changes.forEach(change => {
             const color = GitFileStatus.getColor(change.status, change.staged);
             bucket.set(change.uri, {
                 bubble: true,
-                colorId: color.substring(12, color.length - 1).replace(/-/g, '.'),
+                colorId: this.colorsEnabled ? color.substring(12, color.length - 1).replace(/-/g, '.') : undefined,
                 tooltip: GitFileStatus.toString(change.status),
                 letter: GitFileStatus.toAbbreviation(change.status, change.staged)
             });
         });
     }
 
     provideDecorations(uri: URI, token: CancellationToken): Decoration | Promise<Decoration | undefined> | undefined {
-        return this.decorations.get(uri.toString());
+        if (this.decorationsEnabled) {
+            return this.decorations.get(uri.toString());
+        }
```

Comment:
Minor: We should be more explicit about what we return when decorations are disabled:
```suggestion
        if (this.decorationsEnabled) {
            return this.decorations.get(uri.toString());
        }
        return undefined;
```